client.go

```diff
@@ -394,20 +402,72 @@ func (nc *NativeClient) Session(timeout time.Duration) (*ssh.Session, *SessionIn
 	return session, sessionInfo, nil
 }
 
+func (nc *NativeClient) saveSession(client *ssh.Client, sessionInfo *SessionInfo) {
+	if client == nil && sessionInfo == nil {
+		return
+	}
+	// clear any existing sessions
+	nc.SessionStop()
+	nc.SSHClient = client
+	nc.SessionInfo = sessionInfo
+}
+
+func (nc *NativeClient) SessionStart(timeout time.Duration) error {
+	client, sessionInfo, err := nc.Connect(timeout)
+	if err != nil {
+		return err
+	}
+	nc.saveSession(client, sessionInfo)
+	return nil
+}
+
+func (nc *NativeClient) SessionStop() {
+	if nc.SessionInfo != nil {
+		nc.SessionInfo.CloseAll()
+		nc.SessionInfo = nil
+	}
+	if nc.SSHClient != nil {
+		nc.SSHClient.Close()
+		nc.SSHClient = nil
+	}
+}
+
 // Output returns the output of the command run on the remote host.
 func (client *NativeClient) Output(command string) (string, error) {
 	return client.OutputWithTimeout(command, client.DefaultClientConfig.Timeout)
 }
 
 // Output returns the output of the command run on the remote host.
 func (client *NativeClient) OutputWithTimeout(command string, timeout time.Duration) (string, error) {
-	session, sessionInfo, err := client.Session(timeout)
-	// even on failure, intermediate hop connections must close
+	var sessionInfo *SessionInfo
+	var err error
+
+	sshClient := client.SSHClient
+	if sshClient == nil {
+		// If no ssh client in cache, then start a
+		// new session
+		sshClient, sessionInfo, err = client.Connect(timeout)
+		if err != nil {
+			return "", err
+		}
+		defer sessionInfo.CloseAll()
+		defer sshClient.Close()
+	}
+	session, err := sshClient.NewSession()
 	if err != nil {
-		return "", err
+		if client.SSHClient == nil {
+			return "", err
+		}
+		// if cached client was used, then retry with new client
```

Comment:
When re-trying is where we might end up creating multiple sessions from different threads.